Silence State from API and UI

mxinden opened the pull request:
* Return Silence state on `/silences`
  silenceState = "expired" | "active" | "pending"
  ```
  "status": {
    "state": "expired"
  }
  ```
* Restyle SilenceList view
  * Group by Silence state
  * Adjust styling to AlertList view
* Adjust Silence view

---

![screen shot 2017-05-10 at 16 32 49](https://cloud.githubusercontent.com/assets/7047859/25904139/9d0af654-359e-11e7-9e85-da1553082f2d.png)

Fixes issue #745 

This does not introduce the quick filter via three check-boxes. I would suggest that we don't really need those anymore? What do you think?

mxinden commented:
As this is the last issue in the milestone 4, we could merge the new UI into master once this is merged.
=> No more syncing between `master` and `ui-rewrite`

w0rm commented on ui/app/src/Utils/String.elm:
```diff
@@ -0,0 +1,19 @@
+        Nothing ->
+            ' '
+        Just (char, _) -> char
+
```
Why not use `uncons` directly in `capitalizeFirst`?

```
capitalizeFirst : String -> String
capitalizeFirst string =
    case String.uncons string of
        Nothing ->
            string

        Just ( char, rest ) ->
            String.cons (Char.toUpper char) rest
```

mxinden replied:
Thanks, that looks better.

w0rm commented on ui/app/src/Views/Shared/SilenceBase.elm:
```diff
@@ -63,3 +68,34 @@ matcherButton matcher =
+deleteButton : Silence -> Html Msg
+deleteButton silence =
+    if silence.status.state == "expired" then
+        i [] []
```
Maybe change this to `text ""` or make all buttons return `Maybe (Html Msg)`, then you could render them with `List.filterMap identity maybeButtons`

w0rm commented on ui/app/src/Views/SilenceList/Views.elm:
```diff
@@ -1,62 +1,76 @@
+        silencesView =
             if List.isEmpty silences then
-                div [ class "mt2" ] [ text "no silences found" ]
+                div [ class "" ] [ text "No silences found" ]
```
Does it really need an empty class?

mxinden replied:
Good catch, thanks!

w0rm commented on ui/app/src/Views/SilenceList/Views.elm:
```diff
@@ -1,62 +1,76 @@
+
+filterSilencesByState : List Silence -> String -> List Silence
+filterSilencesByState silences state =
+    List.filter (\silence -> silence.status.state == state) silences
```
May be written shorter `List.filter (\{status} -> status.state == state) silences`

w0rm replied:
Or even shorter `List.filter (.status >> .state >> (==) state) silences`

w0rm replied:
Or, you can change the order of the arguments,  which I think makes sense for the filter function. In this case you don't need to specify `silences`

```
filterSilencesByState: String -> List Silence ->  List Silence
filterSilencesByState state =
    List.filter (\{status} -> status.state == state)
```

stuartnelson3 commented on ui/app/src/Silences/Types.elm:
```diff
@@ -37,6 +44,12 @@ type alias Silence =
+    }
+
+
+type alias SilenceStatus =
```
Since this is already in `module Silences.Types`, I would prefer it just be `Status` to avoid some stutter.

stuartnelson3 replied:
Actually, looking at this, is there a reason it isn't a union type?

```elm
type Status
    = Active
    | Active
    | Pending
```

I think I would prefer this, as it saves us from having to do string matching later to check the status.

mxinden replied:
Haven't thought of that. Sounds good. Changed 👍

stuartnelson3 commented on ui/app/src/Views/SilenceList/Updates.elm:
```diff
@@ -28,13 +28,17 @@ update msg model silence filter =
+                    Debug.log "statusCode" statusCode
+            in
+                -- TODO: "Deleted id: ID" growl
+                -- TODO: Add DELETE to accepted CORS methods in alertmanager
```
`DELETE` has been added, so this can be deleted

stuartnelson3 commented on ui/app/src/Views/SilenceList/Views.elm:
```diff
@@ -1,62 +1,76 @@
+
+filterSilencesByState : String -> List Silence -> List Silence
+filterSilencesByState state =
+    List.filter (\{ status } -> status.state == state)
```
This has a big-O value of `O(n^x)`, where `x == len(states)`. Depending on the number of silences this could be pretty painful. Could you either make a note to replace it with `groupBy` (`O(n)`) in this pr https://github.com/prometheus/alertmanager/pull/775/files#diff-4b0e2e6b80ca6bbf40b7a26e66b231b9R76, or just port it over and make a note to make sure we only have one of them when the above PR gets merged?

mxinden replied:
`groupBy` looks a lot better here, thanks. I will add a note so we can replace it once #775 is merged.

Btw: pretty sure it is `O(x*n)` which equals to `O(n)`, but still a good improvement 👍

mxinden commented:
@stuartnelson3 @w0rm I have addressed all of your comments, thanks for the review! Any other wanted changes?

@fabxc Do you agree with the changes on the `golang` side?

w0rm commented on ui/app/src/Utils/String.elm:
```diff
@@ -0,0 +1,14 @@
+module Utils.String exposing (capitalizeFirst)
+
+import String exposing (left, length, slice, cons, uncons, trim)
```
If you're using `String.uncons` etc.,  then there is no need to import it in the scope.

mxinden replied:
Addressed, thanks.

fabxc commented:
Yea, we can roll with this for now. At some point we might want to at least add a toggle/hide button to each group.

mxinden commented:
@stuartnelson3 Triggered [Circle CI](https://circleci.com/gh/prometheus/alertmanager/749)

@all Thanks for the review.
